storage/src/main/java/org/apache/kafka/storage/internals/log/LogAppendInfo.java

```diff
@@ -0,0 +1,320 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.storage.internals.log;
+
+import org.apache.kafka.common.record.CompressionType;
+import org.apache.kafka.common.record.RecordBatch;
+import org.apache.kafka.common.record.RecordConversionStats;
+import org.apache.kafka.common.requests.ProduceResponse.RecordError;
+
+import java.util.Collections;
+import java.util.List;
+import java.util.Optional;
+import java.util.OptionalInt;
+
+/**
+ * Struct to hold various quantities we compute about each message set before appending to the log.
+ */
+public class LogAppendInfo {
+
+    public static final LogAppendInfo UNKNOWN_LOG_APPEND_INFO = new LogAppendInfo(Optional.empty(), -1, OptionalInt.empty(),
+            RecordBatch.NO_TIMESTAMP, -1L, RecordBatch.NO_TIMESTAMP, -1L,
+            RecordConversionStats.EMPTY, CompressionType.NONE, CompressionType.NONE, -1, -1,
+            false, -1L);
+
+    private Optional<LogOffsetMetadata> firstOffset;
+    private long lastOffset;
+    private OptionalInt lastLeaderEpoch;
+    private long maxTimestamp;
+    private long offsetOfMaxTimestamp;
+    private long logAppendTime;
+    private long logStartOffset;
+    private RecordConversionStats recordConversionStats;
+    private CompressionType sourceCompression;
+    private CompressionType targetCompression;
+    private int shallowCount;
+    private int validBytes;
+    private boolean offsetsMonotonic;
+    private long lastOffsetOfFirstBatch;
+    private List<RecordError> recordErrors;
+    private String errorMessage;
+    private LeaderHwChange leaderHwChange;
```

Comment:
Why are all of these mutable here? In the Scala version, only some of the fields had a `var`.